### ISSUE TITLE
Feature Idea - Component-API

### ISSUE THREAD

**Issue by KonoMaxi** [hotwired/stimulus#581]
# Feature Idea - Stimulus Component API
## Executive Summary
Design an API to support mounting of Javascript-Components for very interactive parts of the application WITHOUT compromising the features of Turbo.
There's a demo for vue, react and svelte adapters I built: https://stimulus-component-api.konow.ski/

## Introduction
Hotwire allows to implement an SPAish user-experience while only writing server-side rendered HTML, which is cool and efficient. But sometimes I struggle with parts of the UI that are just SUPER-interactive and require way too much custom JS for my taste.

Some examples are:
- maps
  - how to structure the code so it doesn't look like my kitchen when I cook spaghetti?
- multi-step wizards - optionally with alternative routes
  - Where to put the state? Put it in session, DOM-attributes, a form or keep it in JS?
  - How to handle step-back and -forward? Will it reset or safe the results?
  - When, how and where to invalidate the state in case some external parameter is changed?

JS-Frameworks like Vue have very clean ways to structure use-cases like this, for example the vue-leaflet plugin. https://codesandbox.io/s/dd0kit?file=/App.vue:716-1653

But once introducing a JS-Framework other Problems arise:
- Where is my state now? JS or DOM?
- Turbo dynamically injects turbo-frames or -streams. How to know when to initialize and UNinitialize these components?
- Once I'm in Vue- or React-Land, how do I find my way back into the Hotwire ecosystem? (i.e. handing page-flow back to turbo)

I tried solving the before mentioned problems with StimulusJS as it provides lifecycle callbacks, has the Value-API for state-management and integrates nicely with Turbo. I found a few "mounting xxx-JS components with StimulusJS" blog posts:
https://www.davedkg.com/stimulusjs-and-mounting-a-react-app or
https://gist.github.com/ryenski/c9de022dc94c16131971b64d49c0778d

I basically copy pasted the identical 20 LOC snippets to a hand full of stimulus-controllers the achieve my goal, but I came to the conclusion that "this code is too WET, I need to get it DRY".

## Prototype
To not have to repeat the "mounting"-logic every time I tried to come up with a nice and **unified** way to mount components to DOM and have their State managed in an "Adapter"-Style. To come up with a more generic approach, I tried to write such an adapter for Vue, React and Svelte - trying to keep the interface consistent.

All adapters implemented have 3 component-lifecycle stages, 3 component functions and 2 framework-references.

Lifecycle Stages:
- create
- mount
- unmount

Functions:
- change:
  - the component communicates state-changes bottom-up
  - the stimulus-controller communicates state-changes top-down
- action:
  - the component can call stimulus-controller actions
- setProperty:
  - the stimulus-controller sets some property on the component that's not managed by the Value-API

Framework references:
The adapters should NOT include any React, Vue or Svelte specific code. So the developer has to pass references to integral framework-parts. 
- setRenderFunction:
  - Vue3 and react need to know how to put HTML in the DOM
- setFactory:
  - Vue2 and React need to know how to create a new Component instance

I published a working demo that shows the results.
https://stimulus-component-api.konow.ski/
Please also have a look at the git-repo. It needs some more refactoring, but it's in a presentable state.

## Required functionality
StimulusJS would need a new API that allows to
1. register component-handlers and
2. register framework-components

### Example:
app.js
```
import { ComponentRegistry } from "@hotwired/stimulus"
import { ReactComponent } from "stimulus-react-adapter"

import { createRoot } from 'react-dom/client'
import { createElement } from 'react'

ReactComponent.setFactory(createRoot)
ReactComponent.setRenderFunction(createElement)

ComponentRegistry.register('react', ReactComponent)
```

my_controller.js
```
import HelloComponent from "../react/HelloComponent"

export default class extends Controller {
  static components = [{
    type: 'react', target: 'mountpoint', component: HelloComponent
  }]
  static values = { text: String, counter: Number }
}
```
And then stimulus automagically mounts `HelloComponent`s to every `mountpoint`-target entering the DOM, using the `ReactComponent`-Adapter and injects the values as react-props.

## The Question
Can you (the maintainers) imagine such a functionality to be part of Hotwire?

I see three ways to implement such a feature:
1. entirely outside the hotwire ecosystem: Just like the popular https://stimulus-use.github.io/stimulus-use/#/ extensions.
  - Create useReact, useVue and useSvelte and anybody can import it.
2. partially inside hotwire ecosystem: component api inside, adapters outside.
  - the api (something like I proposed) is implemented in stimulus core and you can then `yarn add stimulus-react-adapter`
3. All in: Maintaining a component-API and a hand full of popular adapters inside the hotwire ecosystem.

I'd be happy to go for (and contribute to) any approach, but I'd like your Feedback before I produce a (possibly rejected) PR on the stimulus Repository.

**Comment by misterhtmlcss** [comment]
You should just proceed at No.1 now that you see no one has responded. This is a great idea. I think it's too bad no one responded because allowing for some kind of plugin with Stimulus would just allow us devs to make out house bigger and include more people inside the house which is good for rails et al in general. Nice work!

If you decide to chase down option No. 1 please tag me and I'll come check it out. Happy to try contribute.

**Comment by kstratis** [comment]
I'd love to see such thing going through...! 🙏

**Comment by marcoroth** [comment]
Hey @KonoMaxi, thanks for opening this issue and for the work you put into the POCs!

I agree that there are some use-cases where your proposed functionality can make sense and might even be preferable. But I also think that most of the Stimulus users are not going to need that kind of depth for writing Stimulus controllers, also because it kinda speaks against the general ideology behind Stimulus. That's why I think we should start with option 1, before thinking about putting this or part of it into upstream Stimulus.

If there are some general concepts which are missing we can look into a hybrid version between 1 and 2. But I think we should start outside of upstream and see where and how this goes, and also how it's going to be received by the community.

I happy to help or to assist in any way I can. And in case it turns out to be useful in any way: I also worked on a POC a while back, which is kinda similar of what you are proposing here: http://github.com/marcoroth/stimulus-render

**Comment by KonoMaxi** [comment]
Hello @marcoroth,

thanks for the response! After you mentioned it I had a look at your stimulus-render proposal. The general direction is indeed the same, but you went one step further and put actual HTML into the controller. I saw such proposals in the past (your JSX approach looks real nice though), but that's exactly what I tried not to do. I simply wanted to make stimulus the "intermediary" between DOM and JS-Components.

But regarding general concepts that I missed... Now that you mention it, there IS one thing that felt incredibly dirty while building the stimulus-component-api POC: https://github.com/KonoMaxi/stimulus-component-api/blob/7597479c2e7874dfcf948cbddd4b05ac47bbd22d/src/stimulus-component/core/mountHelper.js#L39
In this function I'm creating a proxy to sniff out changes to the valueObserver. Did I oversee some cool "changeCallbackRegistry" in the stimulusJS source-code, or is this currently the way to go? Any help or suggestions are much appreciated :-)

**Comment by trueinviso** [comment]
I believe that there is one small change that could make state management much easier with StimulusJS and turbo.  Unless I'm mistaken, the values API has to be defined on the same element as the controller, it wouldn't work any other way for me:

```
<div data-controller="maps" data-maps-latlng-value=<%= latlng %>
  ...
</div>
```

If we could define the value on an inner div, then the stimulus controller could listen for changes with the [value]Changed function and trigger an action.  This would make it easier to mimic "state" with turbo streams that would update the data value on some action.

Or is the Stimulus convention to just remount the controller on every turbo stream request?

**Comment by jaredcwhite** [comment]
Just throwing this out there:

A conversation around "component APIs" which doesn't start with custom elements (aka web components) in ~2023 is missing the forest for the trees. I would love to see Stimulus fully embrace web components. Right now they can certainly co-exist but there are rough edges and if you introduce shadow DOM stuff falls apart.

I also think as "modest" as Stimulus is in terms of project goals, it's already veering into the territory I believe is now better served by web components. In other words, every Stimulus controller should ask itself if it should actually be a custom element IMHO. I see Stimulus as a useful way to introduce generic behavior "mixins" to existing server-rendered HTML, but when it wades into "this is a widget" territory I question why one would write `<div data-controller="my-widget">` instead of `<my-widget>`. What would be REALLY cool is if you could write `<my-widget>` and then pull some Stimulus guts into the custom element. Now we're talking.

**Comment by KonoMaxi** [comment]
Hey @jaredcwhite, indeed I also use Web Components for some functionality I previously tackled with stimulusControllers - Google's LitElement is quite handy for this stuff and plays nice with import-maps. For me a stimulus-controller is a piece of "life-cycled" javascript that overarches sometimes huge sections of my HTML-pages... Something I really don't want to do with web-components, as I tend to use them for small reusable components that I "sprinkle" all over the place.

My "component-api" angle is more in the direction of "How to put an entire sub-app into my DOM and have it work in an easily configurable way that both dev and customer don't notice they just switched frontend-technology paradigms". I understand this is also possible with web-components, but they sadly have a lot of restrictions - maybe that's why frameworks like svelte have gained popularity this quickly in recent years.

To be precise: This component-API I'm talking about should be able to mount a small react-/vue-App (yes, an app - not a button-webcomponent) into a view, play nice with turbo-hotwire and allow me to persist state via the value-API.

**Comment by jaredcwhite** [comment]
@KonoMaxi So would you say you're describing an "island" architecture then? (aka https://jasonformat.com/islands-architecture/) I've seen a lot of different takes on that. There's even a web component for it! https://github.com/11ty/is-land

I'm just not sure how that would dovetail with Turbo. Unless you wrote some custom Turbo Stream actions, Turbo doesn't naturally provide state updates to client-side frameworks.

**Comment by KonoMaxi** [comment]
Maybe "island architecture" goes into the right direction, but it feels a bit too heavy right now as an approach - maybe after maturing a bit further I'll give it a try :-)

Regarding Turbo integration - I'd not drive it that far. When a turbostream yanks in a `<div data-controller="test"><div data-test-target="mountpoint"></div></div>`, I just want it to use the mountpointTarget as a component root. When the target/controller is deleted it should be gracefully unmounted. Nothing fancy there. For the rest we've got JSON and HTML-Forms :-)

As I got a rails project in maintenance where I mounted a hand full of Vue2 components (just 3 or 4 actually), I finally decided to build the `stimulus-component-api`-package... Now my code looks a bit cleaner and maybe others will enjoy it, too.

In case someone wants to fiddle around with it: Check out the [NPM-Package](https://www.npmjs.com/package/stimulus-component-api) or dive into the [source](https://github.com/KonoMaxi/stimulus-component-api)
